enhancements/authentication/invalid-certs.md

```diff
@@ -0,0 +1,152 @@
+---
+
+# Detecting invalid CA certificates
+
+## Summary
+
+OpenShift 4.10 is going to be rebased against Kubernetes 1.23. This requires using Go 1.17.
+However, starting with Go 1.17 support for legacy invalid is going to be removed, see https://go.dev/doc/go1.17.
+More formally, the temporary `GODEBUG=x509ignoreCN=0` flag has been removed.
+This implies that starting from OpenShift 4.10 invalid certificates will not be trusted any more
+as they will fail verification.
+
+This enhancement suggests a solution to prevent broken workloads from OpenShift 4.9 to OpenShift 4.10
+if invalid certificates are detected.
+
+This enhancment therefore proposes:
+- a way to detect invalid certificates in kube-apiserver and oauth-server
+- a way to prevent upgrades from OpenShift 4.9 to OpenShift 4.10 in the face of invalid certificates
+
+## Motivation
+
+We want to prevent broken workloads due to invalid certificates
+when upgrading from OpenShift 4.9 to OpenShift 4.10.
+
+### Goals
+
+A means to detect invalid certificates for TLS connections between:
+- kube-apiserver and API webhooks
+- kube-apiserver and aggregated API endpoints
+- oauth-server and its external route
+- oauth-server and external identity providers
+
+### Non-Goals
+
+Other core workloads are out of scope for this enhancement.
+
+## Proposal
+
+### User Stories
+
+1. As a cluster admin I want to be prevented from upgrading OpenShift 4.9 to OpenShift 4.10
+if I configured invalid certificates.
+
+2. As a cluster admin I want to get feedback in OpenShift 4.9 if invalid certificates are being detected.
+
+### API Extensions
+
+N/A
+
+### Detecting invalid certificates
+
+#### kube-apiserver
+
+In Kubernetes detection of invalid certificates has been added as part of https://github.com/kubernetes/kubernetes/pull/95396.
+The introduced Prometheus `apiserver_webhooks_x509_missing_san_total` and `apiserver_kube_aggregator_x509_missing_san_total` metrics provides the means to detect invalid certificates against API webhooks and aggregated API endpoints.
+
+To detect invalid certificates we propose to add a new controller in cluster-kube-apiserver that:
+
+1. Regularly executes queries against the cluster internal 
+
+#### oauth-server
+
+For oauth-server we propose the introduction of the following new metric:
+
+- `openshift_auth_external_x509_missing_san_total`: a metric capturing the count of invalid certificates
+presented by an external identity provider.
+
+In case of having invalid custom certificates configured for the external OAuth route
+we propose to extend the existing `RouterCertsDomainValidationController`.
+Formally we propose to add additional logic in `validateRouterCertificates` https://github.com/openshift/cluster-authentication-operator/blob/ff156ab2bdfbdd68b49d76547fea5ec28d9c3639/pkg/controllers/routercerts/controller.go#L129.
+
+#### library-go
+
+The above mentioned logic to validate a `*x509.Certificate` will be provided as a library-go function.
```

Comment:
Which logic? What does it do? Maybe extend it in this section a bit.

Reply:
fixed